challenge_15/action_view.rb

```diff
@@ -25,41 +25,47 @@ class Blog < ActiveRecord::Base; end
 
 class AppController < ActionController::Base
   def root
-    response_body = ""
-    response_body += "<p><strong>Submit a new Blog Post!</p></strong>"
-    response_body += "<form method='post' enctype='application/x-www-form-urlencoded' action='/create-post'>"
-    response_body += "<p><label>Blog Title: <input name='title'></label></p>"
-    response_body += "<p><label>Content: <textarea name='content'></textarea></label></p>"
-    response_body += "<p><button>Submit post</button></p>"
-    response_body += "</form>"
-    render html: response_body.html_safe
+    erb_response = <<~ERB
+      <p><strong>Submit a new Blog Post!</p></strong>
+      <form method='post' enctype='application/x-www-form-urlencoded' action='/create-post'>
+      <label>Blog Title: <input name='title'></label></p>
+      <p><label>Content: <textarea name='content'></textarea></label></p>
+      <p><button>Submit post</button></p>
+      </form>
+    ERB
+    render inline: erb_response
   end
 
   def show_data
-    response_body = ""
-    response_body += "<ul>"
+    @blogs = Blog.all
 
-    Blog.all.each do |blog|
-      response_body += "<li>"
-      response_body += "<strong>Title: #{CGI.escape_html(blog.title)}</strong>, Content: #{CGI.escape_html(blog.content)}"
-      response_body += "</li>"
-    end
-    response_body += "</ul>"
-    render html: response_body.html_safe
+    erb_response = <<~ERB
+      <ul>
+      <% @blogs.each do |blog| %>
```

Comment:
WOW, this was quite the journey! 😂  Thanks for writing this up, I hadn't thought about unveiling the magic of how controller ivars are made available to the views 🤯  

Kind of random side question - what's the purpose of slicing off the `@` when storing the controller's ivars in a hash, if only to add the `@` back on when doing the instance variable set on the view context? Is there a reason we can't store `'@blogs'` directly in the hash?

Reply:
Extremely good question! I don’t know the full answer to this, but it’s at least in part because `#render` supports [an `:assigns` option](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/rendering.rb#L110). If an `:assigns` hash is provided, it gets [passed into `#assign`](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/rendering.rb#L113) just like the hash that comes out of `#view_assigns`, which results in additional instance variables being defined on the view context. So the style of keys in that hash is part of the public API of `#render`, and it’s aesthetically nicer to be able to say `assigns: { more_blogs: MoreBlog.all }` than `assigns: { '@more_blogs' => MoreBlog.all }` in your `render` call, hiding the messy detail that these values are going to end up as instance variables.

So in summary: by slicing off the `@` when automatically populating the `#view_assigns` hash, we make it look more like a hash that a human would write in their code, which makes it easier to incorporate actual human-written hashes into the same mechanism. If the hash was purely a private implementation detail then I don’t think there’d be any point in doing the back-and-forth conversion.

Reply:
Ah neat!! Makes total sense to do this in the interest of providing a nicer API for `#render`. Thanks for the context! ❤️